### PR TITLE
KAFKA-8122: Fix Kafka Streams EOS integration test

### PR DESCRIPTION
**Pull request by mjsax**
We tried to use `commitRequested` to synchronize the test progress, however, this mechanism seems to be broken. Note that `context.commit();` is a request that KS should commit asap -- however, after `context.commit()` returned only an internal flag is set and the commit is not executed yet.

Hence, after the counter is increased to 2, there is no guarantee what happens next: either we commit, or we might actually `poll()` for new data and if `writeInputData(uncommittedDataBeforeFailure);` executed before KS `poll()` again, we might process more than 10 records per partition before we actually commit, and hence the test fails.

A better way seems to be, to read the committed output data before writing new data to make sure the new data is not part of the first transactions and stays uncommitted before we inject the error.

Call for review @guozhangwang @ableegoldman @abbccdda @cpettitt-confluent

### CODE REVIEW

**Comment by guozhangwang** [comment]
Not sure if it is really the root cause, since in StreamThread:

```
for (int i = 0; i < numIterations; i++) {
                    processed = taskManager.process(now);

                    if (processed > 0) {
                        final long processLatency = advanceNowAndComputeLatency();
                        processSensor.record(processLatency / (double) processed, now);

                        // commit any tasks that have requested a commit
                        final int committed = taskManager.maybeCommitActiveTasksPerUserRequested();

                        if (committed > 0) {
                            final long commitLatency = advanceNowAndComputeLatency();
                            commitSensor.record(commitLatency / (double) committed, now);
                        }
                    } else {
                        // if there is no records to be processed, exit immediately
                        break;
                    }
                }
```

That is, if context.committed is called, we should NOT process any more records for that task any more.

**Comment by guozhangwang** [comment]
A "wild" idea about the possible root cause here: the streams instance is configured with two threads, so if one of the threads joins group very, very late, the other thread would own both tasks (as we have two partitions for the input topic), process the first batch of 10 records for both tasks, commit, and then process 5 more records. AND at this time if the other thread finally joined, it would cause a normal rebalance and the tasks committing their state after 15 records processed -- in other words, it would commit once at 10 due to `context.commit`, but then commit again at 15 due to rebalance.

We can exclude that possibility by waiting on the registered state listener to transit to `RUNNING`, at which the stream threads would be all rebalanced and running. And we'd better add this check for other test cases that rely on the assumption of committing behaviors as well, e.g. `shouldNotViolateEosIfOneTaskFails`.

**Comment by mjsax** [comment]
Good point -- I'll update the PR accordingly.

**Comment by mjsax** [comment]
@guozhangwang Updated this PR.

**Review by guozhangwang — APPROVED** [review]

**Comment by mjsax** [comment]
Java 11 / 2.12 failed -- result are not available any longer.
Java 11 / 2.13 and Java 8 passed.

Retest this please.

Call for second review @ableegoldman @abbccdda @cpettitt-confluent

File: streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java
Comment (cpettitt-confluent) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java]: Maybe worth lifting to TestUtils since this pattern is showing up a lot?

---

Minor nit: if you use a CountDownLatch then you don't need to poll and thus can respond a bit more quickly to completion. I think the default resolution on wait is 100ms so it's not a huge difference, but it also seems a bit simpler to do a timed wait on CDL? Your choice.

**Review by cpettitt-confluent — APPROVED** [review]
It would be nice to pull the wait for start into TestUtil or similar since it is useful in several places.

File: streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java]: Ack.

Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java]: Ack

**Comment by mjsax** [comment]
I also include a follow up to #7476 to use the new test-utils methods.

**Comment by cpettitt-confluent** [comment]
I also need that util for one of my tests, so I'll be keeping an eye on this ticket. Thanks!

File: streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java
Comment (cpettitt-confluent) [inline streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java]: The error message is not accurate because it indicates that the wait was for timeoutMs, but it was interrupted so it is quite likely that the wait was less. As we don't expect an interrupt anyway rethrowing it seems the easiest way to handle this?

File: streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java
Comment (cpettitt-confluent) [inline streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java]: Should be something like:

```
assertThat("KafkaStreams did not transit to RUNNING state within " + timeoutMs + " milli seconds.",
    countDownLatch.await(timeoutMs, TimeUnit.MILLISECONDS));
```

File: streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java]: +1

**Comment by guozhangwang** [comment]
LGTM @mjsax after you've addressed the latest comment please feel free to merge.

**Review by bbejeck — APPROVED** [review]

**Comment by mjsax** [comment]
Updated. If no objection raised, I'll merge after Jenkins passed.

**Comment by cpettitt-confluent** [comment]
LGTM!

**Comment by mjsax** [comment]
Cherry-picked to `2.4`, branch.

**Pull request merged**
